[indylambda] Improve test coverage

retronym opened the pull request:
Adding tests for the selective use of boxing bridge methods
and to show that specialization is not subverted by indylambda.

Other aspects of indylambda are tested by tests like:
- run/lambda-serialization.scala
- run/indylambda-boxing

When those tests were written, they only tested the old backend.
However, now that we have Java 8 and the new backend avaialble
by default to partest, they provide the intended coverage.

retronym committed:
[indylambda] Improve test coverage

Adding tests for the selective use of boxing bridge methods
and to show that specialization is not subverted by indylambda.

Other aspects of indylambda are tested by tests like:

  - run/lambda-serialization.scala
  - run/indylambda-boxing

When those tests were written, they only tested the old backend.
However, now that we have Java 8 and the new backend avaialble
by default to partest, they provide the intended coverage.

retronym commented:
Review by @SethTisue or @lrytz

lrytz commented:
see also https://github.com/scala/scala/pull/4603

lrytz commented on test/junit/scala/tools/nsc/backend/jvm/IndyLambdaTest.scala:
```diff
@@ -0,0 +1,54 @@
+  }
+}
+
+class IndyLambdaTest {
```
You need some more boilerplate: the class needs to extend `ClearAfterClass`, and `ClearAfterClass.stateToClear` needs to be set. Maybe this can be improved..

```
class IndyLambdaTest extends ClearAfterClass {
  ClearAfterClass.stateToClear = IndyLambdaTest
  ...
```

lrytz commented on test/junit/scala/tools/nsc/backend/jvm/IndyLambdaTest.scala:
```diff
@@ -0,0 +1,54 @@
+import scala.collection.JavaConverters._
+
+object IndyLambdaTest extends ClearAfterClass.Clearable {
+  var compiler = newCompiler(extraArgs = "-Ybackend:GenBCode")
```
`extraArgs` is redundant

lrytz commented on test/junit/scala/tools/nsc/backend/jvm/IndyLambdaTest.scala:
```diff
@@ -0,0 +1,54 @@
+    // This is because Scala's unboxing of null values gives zero, whereas Java's throw a NPE.
+
+    // 1. Here we show that we are calling the boxing bridge (the lambda bodies here are compiled into
+    //    methods of `(I)java/lang/Object;` / `(I)java/lang/Object;` respectively.)
```
comment should say `(I)Ljava/lang/Object; / (Ljava/lang/Object;)I`

lrytz commented:
otherwise, LGTM

adriaanm commented:
(admin note: please don't start your comment with LGTM until it's actually good to merge -- at some point the bot will start merging PRs as soon as the reviewer signs off, so may as well get into this habit now)

adriaanm commented:
I'm going to push this to M3, since all other M2 PRs have been merged and this can wait.

retronym committed:
Address review feedback

  - Add requisite boilerplate for test cleanup to IndyLambdaTest
  - ... and to the spot where I copy/pasted the wrong code from
  - Fix comment.

retronym commented:
@lrytz I've addressed your comments.

lrytz commented:
LGTM
